src/Sarif.Viewer.VisualStudio/Sarif/Location.Extensions.cs

```diff
@@ -40,5 +41,61 @@ public static LocationModel ToLocationModel(this Location location, Run run, int
 
             return model;
         }
+
+        public static string ExtractSnippet(this Location location, Run run)
+        {
+            var physicalLocation = location.PhysicalLocation;
+            var artifactLocation = location.PhysicalLocation?.ArtifactLocation;
+            var region = location.PhysicalLocation?.Region;
+            var uri = location.PhysicalLocation?.ArtifactLocation?.Uri;
+
+            if (uri == null || region == null || region.IsBinaryRegion || physicalLocation == null)
+            {
+                return string.Empty;
+            }
+
+            if (region.Snippet != null)
+            {
+                return region.Snippet.Text;
+            }
+
+            if (artifactLocation.Uri == null && artifactLocation.Index >= 0)
+            {
+                // Uri is not stored at result level, but we have an index to go look in run.Artifacts
+                // we must pick the ArtifactLocation details from run.artifacts array
+                Artifact artifactFromRun = run.Artifacts[artifactLocation.Index];
+                artifactLocation = artifactFromRun.Location;
+            }
+
+            // If we can resolve a file location to a newly constructed
+            // absolute URI, we will prefer that
+            if (!artifactLocation.TryReconstructAbsoluteUri(run.OriginalUriBaseIds, out Uri resolvedUri))
+            {
+                resolvedUri = artifactLocation.Uri;
+            }
+
+            if (!resolvedUri.IsAbsoluteUri)
+            {
+                return string.Empty;
+            }
+
+            FileRegionsCache fileRegionsCache = new FileRegionsCache();
```

Comment:
>fileRegionsCache [](start = 29, length = 16)

should we set this as static variable, since this is a cache?

Reply:
This is a extension methods class, not a good place to store static cache.
I updated code to use the FileRegionsCache from parameter, which get from CodeAnalysisResultManager.Instance.RunIndexToRunDataCache. Similar to how its used in SarifErrorListItem.cs. What do you think?

---
In reply to: [553624644](https://github.com/microsoft/sarif-visualstudio-extension/pull/319#discussion_r553624644) [](ancestors = 553624644)

Reply:
makes sense! thank you for adding that!

---
In reply to: [553735607](https://github.com/microsoft/sarif-visualstudio-extension/pull/319#discussion_r553735607) [](ancestors = 553735607,553624644)